### PR TITLE
Fix #96: consider superclass

### LINKED ISSUE #96: Fail to recognize Generics in Superclass
**Input:**
1.java
```java
package com.test;

public class TestClass extends SuperClass<Old> {

}

```
2.java
```java
package com.test;

public class TestClass extends SuperClass<New> {

}

```
**Output:**
```bash
java -jar gumtree-spoon-ast-diff-1.17-jar-with-dependencies.jar 1.java 2.java
no AST change
```

I would have expected the change "Old > New" to be represented in the AST.
This is happening because this is a `CtTypeReference` and all `CtReference`s are ignored.

### PR DESCRIPTION
**Pull request by sedflix**
#96 
This PR allows us to consider extends arguments//superclass and introduces the corresponding test cases.

### CODE REVIEW

File: src/test/java/gumtree/spoon/AstComparatorTest.java [hunk 1]
```diff
@@ -1703,10 +1687,10 @@ public void testModifEmpty() throws Exception {
 		List<Operation> actions = resulta.getRootOperations();
 		resulta.debugInformation();
 
-		assertEquals(1, actions.size());
+		assertEquals(2, actions.size());
```
Comment (monperrus) [inline src/test/java/gumtree/spoon/AstComparatorTest.java]: same here

**Comment by sedflix** [comment]
At the current moment, it is able to identify `new Callback()` -> `new Callback<T>()`. But it can't identidy `Response` -> `Response<T>`.

**Comment by sedflix** [comment]
Given that we are considering generics, how many actions should be present in the following snippet ideally?

```
CtClass c1a = Launcher.parseClass(" class BehaviorCall implements Call{\n"
				+ "final AtomicReference failureRef = new AtomicReference<>();\n"
				+ "final CountDownLatch latch = new CountDownLatch(1);\n" + "\n" + " enqueue(new Callback<T>() {\n"
				+ "  @Override public void onResponse(Response<T> response) {\n" + "     responseRef.set(response);\n"
				+ "     latch.countDown();\n" + "   }\n" + "}\n" + ")\n" + "\n" + "}");

CtClass c2a = Launcher.parseClass("class BehaviorCall implements Call {\n"
				+ "final AtomicReference failureRef = new AtomicReference<>();\n"
				+ "final CountDownLatch latch = new CountDownLatch(1);\n" + "enqueue(new Callback() {\n"
				// Here the difference
				+ "@override public void onResponse(Call call, Response response) {\n" + "responseRef.set(response);\n"
				+ "latch.countDown();\n" + "}\n" + "}\n" + ")\n" + "}");
```

**Comment by monperrus** [comment]
it depends whether the parameter change is considered as a move or as an insert.

in general, there is no unique answer, it also depends on the tree encoding.

what's the current behavior.

**Comment by sedflix** [comment]
The current behavior only considers cases of generics in constructor calls(example: `new Callback<T>()`) and extends' argument(example: `extends SuperClass<Wow>`). This happens as the label assigned to the constructor call or the superclass in extends' argument is done using `toString()` which also contains generic part also(as compared to just the class name). 
That's why at the current momemt, it is able to identify new `Callback() -> new Callback<T>()` but not `Response -> Response<T>`.

Right now, a check like below would consider all the generics(but most probably some more cases):
```
  if (element instanceof CtTypeReference
                && ( element.getRoleInParent() == CtRole.SUPER_TYPE // Checks super classs
                || (element.getRoleInParent() == CtRole.TYPE_ARGUMENT // Checks Generic 
                    && (element.getParent().getRoleInParent() == CtRole.TYPE   // Checks generic in other cases
                          || element.getParent().getRoleInParent() == CtRole.SUPER_TYPE)  // Checks Generic in extends' case
                    && ((CtTypeReference)element.getParent()).getActualTypeArguments().get(0) == element)))
```

The above conditions lead to a weird failure of one test case, ie: DiffTest.testToString().
image

**Comment by sedflix** [comment]
@monperrus I just wanted to confirm if this will be considered as a bug in Spoon.

For the following code:
```
enqueue() {
        new Callback<T>() {
            @Override
            public void onResponse(Response<T> response) {
                responseRef.set(response);
                latch.countDown();
            }
  };
```
The Spoon meta-model looks like below:
image

The area of interest is highlighted in the image.

File: src/test/java/gumtree/spoon/AstComparatorTest.java [hunk 2]
```diff
@@ -1541,8 +1527,9 @@ public void test_issue201812() throws Exception {
 		List<Operation> actions = result.getRootOperations();
 		result.debugInformation();
 
-		assertEquals(1, actions.size());
+		assertEquals(2, actions.size());
 		assertTrue(result.containsOperation(OperationKind.Insert, "Parameter", "call"));
+        assertTrue(result.containsOperation(OperationKind.Update, "TypeReference", "Callback<T>"));
```
Comment (monperrus) [inline src/test/java/gumtree/spoon/AstComparatorTest.java]: the type of this node should be "SUPER_TYPE" and not "TypeReference". could you have a look?

**Comment by monperrus** [comment]
OK, this is not a bug in Spoon, it's only a bug in the Gumtree encoding. the anonymous class (nested
type) should be in the Gumtree model.

File: src/test/java/gumtree/spoon/AstComparatorTest.java (hunk 2 above)
Comment (sedflix) [reply src/test/java/gumtree/spoon/AstComparatorTest.java]: Just to confirm, the above comment means that we need to create a new node type called SuperType which will be a subclass of TypeReference?

Are all the node types in this repo a subset of node types in Spoon?
If not, can you point me to implementation of such a Node?
If yes, I may need some help in figuring this out.

Comment (monperrus) [reply src/test/java/gumtree/spoon/AstComparatorTest.java]: > we need to create a new node type called SuperType 

Yes, a new Gumtree node type called `SUPER_TYPE`

> which will be a subclass of TypeReference?

No. In Gumtree nodes, there is no "subclass" relation ship

> Are all the node types in this repo a subset of node types in Spoon?

No. Some Gumtree nodes have no corresponding Spoon node types (eg `Modifiers`). `SUPER_TYPE` will be one of them.

File: src/main/java/gumtree/spoon/builder/NodeCreator.java [hunk 3]
```diff
@@ -95,4 +96,19 @@ private String getClassName(String simpleName) {
 		super.visitCtMethod(e);
 	}
 
+    @Override
+    public void scanCtReference(CtReference reference) {
+
+
+        if (reference instanceof CtTypeReference && reference.getRoleInParent() == CtRole.SUPER_TYPE) {
+            ITree superType = builder.createNode("SUPER_TYPE", reference.toString());
+            CtWrapper<CtReference> k = new CtWrapper<CtReference>(reference, reference.getParent());
```
Comment (sedflix) [inline src/main/java/gumtree/spoon/builder/NodeCreator.java]: Is this the right way to use CtWrapper?

**Comment by monperrus** [comment]
Looks good to me, it seems we're done!
One more thing: could you remove the changes in LabelFinder.java?

**Comment by monperrus** [comment]
Thanks a lot @sedflix!

**Pull request merged**